<!DOCTYPE html>
<html>

<head>
  <link rel="stylesheet" type="text/css" href="post.css">
  <title>Charlotte Merzbacher</title>
  <meta name="description" content="Charlotte Merzbacher is Production Engineer from San Francisco, C.A.">
  <meta name="author" content="Charlotte Merzbacher">
  <meta charset="utf-8">
</head>

<body>
  <div id="homepage">
    <div class="container">
      <div class="small">
        <div class='page-title'>
          <a href='../index.html'>
            <h1>Charlotte Merzbacher</h1></a>
          <a href='../index.html'>
            <h2>BIOTECH AND DATA SCIENCE</h2></a>
        </div>

        <div class="post">
          <div id='post-date'>August 6, 2019</div>
          <h2> FROM BLAST TO BEAST </h2>
          <div id='subtitle'>Building Phylogenetic Trees from Sequence Data</div>
          <p>The NCBI BLAST web service is awesome, but it can be difficult to
            set your exact algorithm parameters and parse the output to use in
            evolutionary analyses. In this tutorial, I’ll be taking you from a
            single sequence of interest all the way through construction of a
            phylogenetic tree using BEAST2. I’m not an expert on this by any
            means, but this is just the tutorial I would have liked to read as I
            started on this project. Please contact me with any feedback.
            (Note: This tutorial is for nucleotide sequences. For protein
            sequences, see the section at the end.)</p>

          <h3>System Requirements</h3>
          <p>I’m running Mac OS Mojave with Python 3 and BEAST2. I use Atom with
            the Hydrogen package so I can easily run lines as in a Jupyter
            notebook.</p>

          <h3>Installing Biopython</h3>
          <p>I’ll be using the <a href="http://biopython.org/DIST/docs/
            tutorial/Tutorial.html#htoc89">Biopython package</a> to parse my
            BLAST results and automate things. Install with:</p>
          <code>> pip install biopython</code>
          <p>And import with:</p>
          <code>from Bio import ... as ...</code>

          <h3>Running BLAST online with Biopython</h3>
          <p>The next step is to create a FASTA file or a string with the
            sequence you want to search in it. Read the file in (or assign the
            string to) a variable. We call qblast() in the Bio.Blast.NCBIWWW
            module, which takes the type of search, the online database, and the
            sequence:</p>
          <pre><code>from Bio.Blast import NCBIWWW
result = NCBIWWW.qblast(“blastn”, “nt”, sequence)</code></pre>
          <p>In this case, we’re running a BLASTn search on the nucleotide
            database. The other abbreviations can be found in the BLAST guide
            (think blastp, tblastx, etc.). You can also change the format of the
            results. The default format is XML, which we will use below. </p>
        </div>

        <h3>Parsing BLAST output</h3>
        <p>Once we have our BLAST results, we save the XML file. This is mostly
          so we have a file to go back to and don’t need to run the BLAST search
           again in case of disaster.</p>
        <pre><code>with open ('results.xml', 'w') as save_file:
    blast_result  = result.read()
    save_file.write(blast_result)</code></pre>
        <p>Then, we will import the SearchIO and SeqIO submodules and use them
          to parse the XML file and create a FASTA file. (Note: There is also a
          BLAST-specific XML parser in biopython, but it creates a BLAST record
          class, which isn’t what we want here.)</p>
          <pre><code>from Bio import SearchIO
from Bio import SeqIO

blast_result = SearchIO.read('results.xml', 'blast-xml')
records = []
for hit in blast_result:
    records.append(hit[0].hit)
    SeqIO.write(records, 'results.fa', 'fasta')</code></pre>

        <h3>Converting using SeqMagick</h3>
        <p>Unfortunately, Biopython doesn’t seem to have an easy way to create a
           Nexus file, which is a tree file for BEAST. Instead, we can use the
           package seqmagick. (<a href='https://seqmagick.readthedocs.io/en/la
           test/'>Documentation here</a>.) To install, just use:</p>
        <code>> pip install seqmagick</code>
        <p>Seqmagick has a helpful convert function which automatically parses
          several file extensions including .fa (FASTA) and .nex (Nexus). For
          Nexus files, the –alphabet flag is needed to specify DNA, RNA, etc.
          We run the following line in Terminal:</p>
        <code>> seqmagick convert --alphabet dna "results.fa" "results.nex"</code>
        <p>If everything goes well, you’ll now have a nice clean nexus file
          like the one below. Make sure all the aligned sequences are the same
          length.</p>

        <h3>Downloading and installing BEAST</h3>
        <p>(Note: Most of this section I learned from the book Bayesian
          Evolutionary Analysis with BEAST 2, by Alexei J. Drummond and Remco R.
           Bouckaert. I’d recommend reading at least the first 5 chapters for
           anyone unfamiliar with Bayesian analysis for trees.)</p>
        <p>BEAST can be downloaded from the website <a href='https://www.beast2.
          org/'>here</a>. The installation is
          simple, but once its complete, make sure you also have the Bayesian
          Evolutionary Analysis Utility, or BEAUti, installed as well.</p>

        <h3>Setting model parameters with BEAUti</h3>
        <p>First, load the Nexus file you created using seqmagick into BEAUti by
           selecting File > Import Alignment. In my case, I had only one
           partition; however, for protein sequences multiple partitions
           (coding/noncoding, etc.) may appear. If you are working with multiple
            sample dates, you can include that information in the Tip Dates tab;
             however, phage and bacterial evolution rates are slow enough that
             we will treat all samples as contemporaneous.</p>
        <p>Select the Site Models tab and set the Gamma Category Count parameter
           to 4 (an appropriate range is usually 4-6). Now select a model from
           the Subst. Model dropdown menu that fits the data. In this case, I’ll
            be using HKY. </p>
        <p>To set the clock model, go to the corresponding tab and select the
          model that best fits your data. If you’re not sure how clock-like the
          rate of evolution is, run the analysis with a random local clock
          model, which considers each branch in the tree with a separate clock
          rate. </p>
        <p>Select the Priors tab. In this case, I will be using a Yule model as
           my data comes from different species. This is a pretty complex
           process which I won’t cover in detail here, but the BEAST book and
           tutorials are helpful!</p>
        <p>The MCMC tab determines the general parameters for the BEAST run. The
           chain length gives the number of steps the Monte Carlo algorithm will
            make before finishing. I set this to 1,000,000 to reduce the run
            time. (Note: If you’re not sure about your prior choices, doing a
            run with a chain length as short as 10,000 can give you an idea of
            the results, especially if you’re running on an older computer.)</p>
        <p>Save the file as “[filename].xml”. We are now ready to run the file
          through BEAST.</p>

        <h3>Running BEAST</h3>
        <p>Congrats, we’ve almost made it to the end of this tutorial! Now just
          open BEAST, select the XML file you created, and hit run. No need to
          change the random number seed etc. A log window will pop up and run
          for a few minutes. Once it’s done, you’ll see a list of parameters
          like this:</p>
        <p>To visualize the tree, you can use DensiTree or FigTree, both of
          which are available online or come with the BEAST download. Just open
          the program and import the .trees file BLAST spit out, and you’ll get
          something like this:</p>
        <p>I like checking my results in DensiTree quickly to see if the
          parameters I chose were reasonable. For more information on
          visualizing trees, see the DensiTree manuals <a href='https://www.cs.
          auckland.ac.nz/~remco/DensiTree/'>here.</a></p>
        <p>I won’t go into the complete analysis of BEAST results, but Tracer
          (<a href='http://tree.bio.ed.ac.uk/software/tracer/'>found here</a>)
          is useful for examining the underlying distributions and
          more specific descriptive statistics about your tree. Good luck!</p>

        <h3>Note on protein sequences</h3>
        <p>Protein sequences can be downloaded from BLASTp or BLASTx results as
          an XML, converted to a FASTA with the script above, and converted into
           a multiple alignment using COBALT (<a href='https://www.ncbi.nlm.nih.
           gov/tools/cobalt/cobalt.cgi'>here</a>). COBALT Nexus files can then
            be directly inputted into BEAUTi. Yes, this is much easier than the
            nucleotide route. Yes, I found this after doing all the above work
            and was very frustrated. But hey! We did some learning along the
            way.</p>

        <div class="credit">
          <p>built by charlotte merzbacher</p>
        </div>

      </div>
    </div>
  </div>
</body>

</html>
